### DISCUSSION TITLE
Custom Task Instructions Framework

### DISCUSSION THREAD

**Discussion by ryantoth2410** [cline/cline#514] [Feature Requests]
# Custom Task Instructions Framework

## Overview
This framework allows users to define custom steps that MUST be followed when performing tasks. These instructions can be set at the main project level or for specific subdirectories.

## Key Components
1. /docs/task-instructions.md: Main file for project-wide task instructions
2. /*/task-instructions.md: Subdirectory-specific task instructions (where * represents any subdirectory)

### 1. Creating Custom Instructions
- In /docs/task-instructions.md or subdirectory /*/task-instructions.md, define steps using this format:

```markdown
# Task Instructions

1. [Step 1 description]
2. [Step 2 description]
3. [Step 3 description]
...

Important Notes:
- [Any critical information or warnings]
```

### 2. Hierarchy of Instructions
- Subdirectory instructions take precedence over main instructions when working in that subdirectory
- Main instructions apply when no subdirectory-specific instructions exist

### 3. Mandatory Compliance
- ALL defined steps MUST be followed for EVERY task
- NO steps may be skipped or modified without explicit user permission

### 4. Verification Process
- Before starting any task, locate and read the applicable task-instructions.md
- Confirm understanding of all steps
- If any step is unclear, STOP and ask the user for clarification

### 5. Execution and Documentation
- Follow each step in order
- Document the completion of each step
- If unable to complete a step, STOP and notify the user immediately

### 6. Feedback Loop
- After task completion, review the effectiveness of the instructions
- Suggest improvements or clarifications to the user if needed

## Critical Note
There is an known issue with custom instructions not being followed consistently. To address this:

- At the start of EVERY task, explicitly state: "I will now follow the custom task instructions as defined in [path to relevant task-instructions.md]"
- Before marking a task as complete, confirm: "I have followed all steps as outlined in the custom task instructions"
- If at any point you realize you've deviated from the instructions, STOP immediately, notify the user, and await further direction

Remember: Strict adherence to user-defined task instructions is CRUCIAL. Failure to follow these custom steps is considered a critical error.

**Comment by d0lwl0b** [comment]
We can jointly implement a "mandatory" standard for LLMs.

see: #5912
